src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,673 @@
+# Copyright 2024-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+import warnings
+from collections import Counter, defaultdict
+from collections.abc import Mapping
+from contextlib import nullcontext
+from copy import deepcopy
+from functools import partial
+from itertools import cycle
+from typing import Dict, Iterable, Optional, Union
+
+import torch
+from tqdm import tqdm
+from transformers.pytorch_utils import Conv1D
+
+from peft.tuners.tuners_utils import _find_minimal_target_modules, check_target_module_exists
+from peft.utils.constants import MIN_TARGET_MODULES_FOR_OPTIMIZATION
+from peft.utils.incremental_pca import IncrementalPCA
+from peft.utils.other import _get_submodules
+
+from .config import LoraConfig
+from .layer import Embedding, LoraLayer, _ConvNd
+
+
+UNSUPPORTED_LORA_MODULES = (Embedding, _ConvNd)
```

Comment:
fair point, the only advantage is having a better error message, I can invert this to `SUPPORTED_LORA_MODULES`

Reply:
just was reminded during testing that the actual reason why I did it this way is to automatically include bnb linear layers. Otherwise I would need to do a conditional import I guess if bnb is available. Would you still say its better to do it this way?

Reply:
No, that's why I mentioned

> Eva would generally work with newly added quantization methods

For me, this was more to verify the reasoning here When new LoRA layers are added that are not supported here, we'll most likely forget to update this constant, but it's not the end of the world. However, I think we can make it a bit more likely no to miss this if we move `UNSUPPORTED_LORA_MODULES` to `utils/constants.py` and rename it to `UNSUPPORTED_EVA_LORA_MODULES` or something like this.

Reply:
moving this to `constants.py` doesn't work unfortunately due to some circular import issue with `PeftConfig`

Reply:
Hmm, any idea how that can be? In the [line above](https://github.com/huggingface/peft/pull/2142/files#diff-36e504fe6dcda50b98764f8961b977a65a976f187bc391385299fba1fc07704bR29), we already import from `peft.utils.constants`, so it's unclear to me how a circular import code arise here.

Reply:
It has something to do with me now needing to import from lora/layer.py. I think the circular dependency goes something like this
- in constants.py we now import from lora/layer.py
- lora/layer.py imports from lora/config.py
- lora/config.py imports from peft/config.py
- peft/config.py imports from utils/other.py
- utils/other.py imports from constants.py

Reply:
Ah yes, that makes sense. Then I wonder if we can define the constant inside of `lora/__init__.py`. Just somewhere more visible than here :)

Reply:
hmm unfortunately this does not fix the issue. Putting the variable in `lora/__init__.py` still raises an error.

Reply:
All right, then let's just keep it as is.